mirage/tar_mirage.ml

```diff
@@ -129,18 +122,97 @@ module Make_KV_RO (BLOCK : Mirage_block_lwt.S) = struct
       (* Don't try to read beyond the end of the archive *)
       let total_size_bytes = Int64.(mul t.info.Mirage_block.size_sectors sector_size) in
       let tmp = Cstruct.sub block 0 (min (Cstruct.len block) Int64.(to_int @@ (sub total_size_bytes (mul start_sector 512L)))) in
-      BLOCK.read t.b start_sector [ tmp ]
-      >>= function
-      | Error _ -> Lwt.fail (Failure (Printf.sprintf "Failed to read %s" key))
-      | Ok () -> Lwt.return (Ok [ Cstruct.sub block (to_int start_padding) (to_int length_bytes) ])
-    end
-
-  let size t key =
-    let key = trim_slash key in
-    if not(StringMap.mem key t.map)
-    then Lwt.return (Error (`Unknown_key key))
-    else begin
-      let hdr, _start_sector = StringMap.find key t.map in
-      Lwt.return (Ok hdr.Tar.Header.file_size)
-    end
+      BLOCK.read t.b start_sector [ tmp ] >|= function
+      | Error b -> Error (`Block b)
+      | Ok () ->
+        let buf = Cstruct.sub block (to_int start_padding) (to_int length_bytes) in
+        Ok (Cstruct.to_string buf)
+
+  let list t key =
+    let r = match get_node t.map key with
+      | Ok (Dict (_, m)) ->
+        Ok (StringMap.fold (fun key value acc ->
+            match value with
+            | Dict _ -> (key, `Dictionary) :: acc
+            | Value _ -> (key, `Value) :: acc)
+            m [])
+      | Ok (Value _) -> Error (`Dictionary_expected key)
+      | Error e -> Error e
+    in
+    Lwt.return r
+
+  let to_day_ps hdr =
+    let ts =
+      match Ptime.Span.of_float_s (Int64.to_float hdr.Tar.Header.mod_time) with
+      | None -> Ptime.epoch
+      | Some span -> match Ptime.add_span Ptime.epoch span with
+        | None -> Ptime.epoch
+        | Some ts -> ts
+    in
+    Ptime.(Span.to_d_ps (to_span ts))
+
+  let last_modified t key =
+    let r = match get_node t.map key with
+      | Ok (Dict (hdr, _)) -> Ok (to_day_ps hdr)
+      | Ok (Value (hdr, _)) -> Ok (to_day_ps hdr)
+      | Error e -> Error e
+    in
+    Lwt.return r
+
+  let digest t key =
+    get t key >|= function
+    | Error e -> Error e
+    | Ok data -> Ok (Digest.string data)
+
+  (* Compare filenames without a leading / or ./ *)
+  let trim_slash x =
```

Comment:
Is this function something that can be replaced with `Fpath.normalize` use?

Reply:
We're using [Mirage_kv.Key](https://mirage.github.io/mirage-kv/mirage-kv/Mirage_kv/Key/index.html), where I can't find such a function. But maybe worth an addition!